src/utils/customHooks.js

```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+
+function useLocalStorageState(
+    key,
+    defaultValue = '',
+    // the = {} fixes the error we would get from destructuring when no argument was passed
+    // Check https://jacobparis.com/blog/destructure-arguments for a detailed explanation
+    {serialize = JSON.stringify, deserialize = JSON.parse} = {},
+  ) {
+    const [state, setState] = React.useState(() => {
+      const valueInLocalStorage = window.localStorage.getItem(key)
+      if (valueInLocalStorage) {
+        // the try/catch is here in case the localStorage value was set before
+        // we had the serialization in place (like we do in previous extra credits)
+        try {
+          return deserialize(valueInLocalStorage)
+        } catch (error) {
+          window.localStorage.removeItem(key)
+        }
+      }
+      return typeof defaultValue === 'function' ? defaultValue() : defaultValue
```

Comment:
The point of that is to enable a hook feature called 'lazy initialization' where the hook uses a function as a default value instead of a value. When a function is passed, the hook is optimized to only call that function one time instead of each render
https://ireadyoulearn.info/2021/05/08/react-hooks-usestate-lazy-loading-synced-updates-and-refs-overview/